set.seed(1) 
myPaths <- .libPaths()
new <- c('/storage/zhangyanxiaoLab/suzhuojie/miniconda3/envs/R4.2/lib/R/library')
myPaths <- c(myPaths, new) 
.libPaths(myPaths)
setwd("/storage/zhangyanxiaoLab/suzhuojie/projects/medulloblastoma/")
library(Seurat)
library(vegan)
library(ggplot2)
library(stringr)
library(numbat)
#导入rawdata
se <- readRDS("data/merge_data/before_doubletfinder/annotation_rds/annotation_final_with_neurons/s09_met_anntation.rds")
df_allele_raw <- read.table("data/rawdata/bam/S09-metastatic/S09-metastatic_allele_counts.tsv.gz", sep = '\t', header = TRUE)

#构建ref
annotation_ref <- as.data.frame(se@meta.data$annotation[which(se@meta.data$annotation %in% c("Asstrocyte","Endothelial","Fibroblast","Microglia","Oligodendrocyte","T_cell"))])
rownames(annotation_ref)<-rownames(se@meta.data)[which(se@meta.data$annotation %in% c("Asstrocyte","Endothelial","Fibroblast","Microglia","Oligodendrocyte","T_cell"))]
barcode_ref<-rownames(annotation_ref)
se_ref<-subset(se,cells = barcode_ref)
count_ref<-se_ref@assays$RNA@counts
colnames(annotation_ref)[1]<-"group"
annotation_ref$cell <- rownames(annotation_ref)
ref = aggregate_counts(count_ref,annotation_ref)

#构建cancer
barcode_cancer<-rownames(se@meta.data)[which(!(se@meta.data$annotation %in% c("Asstrocyte","Endothelial","Fibroblast","Microglia","Oligodendrocyte","T_cell")))]
se_cancer<-subset(se,cells = barcode_cancer)
df_allele_cancer <- subset(df_allele_raw,df_allele_raw$cell %in% barcode_cancer)
count_cancer <- se_cancer@assays$RNA@counts

#For cell line data and high-purity tumors, 
#we recommend running the below procedure to identify and exclude regions of clonal deletions/LOH before running Numbat
bulk = get_bulk(
  count_mat = count_cancer,
  lambdas_ref = ref,
  df_allele = df_allele_cancer,
  gtf = gtf_hg38,
)
segs_loh = bulk %>% detect_clonal_loh(t = 1e-4)

out = run_numbat(
  count_cancer, # gene x cell integer UMI count matrix 
  ref, # reference expression profile, a gene x cell type normalized expression level matrix
  df_allele_cancer, # allele dataframe generated by pileup_and_phase script
  genome = "hg38",
  t = 1e-5,
  ncores = 32,
  plot = TRUE,
  segs_loh = segs_loh,
  out_dir = 'merge_data/before_doubletfinder/numbat/S02-primary2/'
  
)
nb=Numbat$new(out_dir='merge_data/before_doubletfinder/numbat/S02-primary2')
plist=list()

cnv_type=nb$joint_post %>% distinct(seg,cnv_state) %>% {setNames(.$cnv_state,.$seg)}
pagoda<-Pagoda2$new(se@assays$RNA@counts,log.scale=T,n.cores=2)
pagoda$adjustVariance(plot=T,gam.k=10) 
pagoda$calculatePcaReduction(nPcs=50,n.odgenes=3e3) 
pagoda$makeKnnGraph(k=40,type='PCA',center=T,distance='cosine') 
pagoda$getKnnClusters(method=infomap.community,type='PCA')
M <- 30; 
pagoda$getEmbedding(type='PCA',embeddingType = 'largeVis', 
               M=M,perplexity=30,gamma=1/M,alpha=1)
pagoda$getEmbedding(type='PCA',embeddingType='UMAP',perplexity=50,verbose=F,n.cores=30)
muts=c('17b')
plist[['17b']] = pagoda$plotEmbedding(
        alpha=0.8,
        size=1, 
        plot.na = F, 
        colors = nb$joint_post %>%
            filter(seg == mut) %>%
            {setNames(.$p_cnv, .$cell)},
        show.legend = T,
        mark.groups = F,
        plot.theme = theme_bw(),
        title = paste0(mut, '(', cnv_type[muts], ')')
    ) +
    scale_color_gradient2(low = 'royalblue', mid = 'white', high = 'red3', midpoint = 0.5, limits = c(0,1), name = 'Posterior')

chr17_pcnv<-nb$joint_post$p_cnv[which(nb$joint_post$seg=='17b')]
chr17_cell<-nb$joint_post$cell[which(nb$joint_post$seg=='17b')]
p_cnv <- as.data.frame(chr17_pcnv,chr17_cell)
p_cnv <- as.data.frame(chr17_pcnv)
rownames(p_cnv)<-chr17_cell
se_cancer$chr17b <- p_cnv

se_plot <- se_cancer@reductions$umap@cell.embeddings %>% as.data.frame() %>% cbind(chr17b=se_cancer$chr17b)
p<-ggplot(se_plot,aes(x=UMAP_1,y=UMAP_2,color=se_plot$chr17b))+geom_point(size=0.2,alpha=1)+scale_color_gradient(low = "white", high = "black")
ggsave("result/merge_data/before_doubletfinder/numbat/S02-primary/chr17b-2.png",p,width=15,height=10)